File: macos/Webyar/Core/Api/WebAnalytics.swift
import Foundation

// MARK: - Website analytics (`/api/web-analytics/:workspaceId/…`), as the web's SEO → Web Analytics reads it

/// The headline numbers for a date range, and the day-by-day trend.
struct WebAnalyticsOverview: Codable, Hashable, Sendable {
    @Lenient var sessions: Int? = nil
    @Lenient var pageviews: Int? = nil
    @Lenient var avgPagesPerSession: Double? = nil
    @Lenient var uniqueVisitors: Int? = nil
    /// Percentage, 0–100, of sessions with a single page view.
    @Lenient var bounceRate: Double? = nil
    @Lenient var avgVisitDurationSeconds: Double? = nil
    var trend: [WebAnalyticsDay]?
    var topChannels: [WebAnalyticsRow]?
    var topPages: [WebAnalyticsPage]?
    var truncated: Bool?
}

struct WebAnalyticsDay: Codable, Hashable, Sendable {
    /// YYYY-MM-DD.
    var date: String
    @Lenient var sessions: Int? = nil
    @Lenient var pageviews: Int? = nil
}

/// One line of a "group by X, count sessions" report.
struct WebAnalyticsRow: Codable, Hashable, Sendable, Identifiable {
    var key: String
    var label: String?
    @Lenient var sessions: Int? = nil
    @Lenient var pageviews: Int? = nil
    var id: String { key }
}

struct WebAnalyticsPage: Codable, Hashable, Sendable, Identifiable {
    var path: String
    @Lenient var views: Int? = nil
    var id: String { path }
}

struct WebAnalyticsEvent: Codable, Hashable, Sendable, Identifiable {
    var eventName: String
    @Lenient var count: Int? = nil
    @Lenient var uniqueSessions: Int? = nil
    /// 0–1: the share of the range's sessions that fired it.
    @Lenient var conversionRate: Double? = nil
    var id: String { eventName }
}

/// A report's rows, and whether the server stopped short of the whole range.
struct WebAnalyticsRows<Row: Codable & Hashable & Sendable>: Codable, Hashable, Sendable {
    var rows: [Row]?
    var truncated: Bool?
}

private struct LiveVisitorsResponse: Decodable { @Lenient var count: Int? = nil }

extension WebyarAPI {
    private func analyticsPath(_ workspaceId: String, _ tail: String) -> String {
        "/api/web-analytics/\(ApiClient.escape(workspaceId))/\(tail)"
    }

    private func range(_ start: String, _ end: String) -> [(String, String?)] {
        [("startDate", start), ("endDate", end)]
    }

    func analyticsOverview(workspaceId: String, start: String, end: String) async throws -> WebAnalyticsOverview {
        try await client.get(analyticsPath(workspaceId, "overview"), query: range(start, end))
    }

    func analyticsLiveVisitors(workspaceId: String) async throws -> Int {
        let r: LiveVisitorsResponse = try await client.get(analyticsPath(workspaceId, "live-visitors"))
        return r.count ?? 0
    }

    /// channel | source | campaign.
    func analyticsTrafficSources(workspaceId: String, dimension: String, start: String, end: String) async throws -> WebAnalyticsRows<WebAnalyticsRow> {
        try await client.get(analyticsPath(workspaceId, "traffic-sources"), query: [("dimension", dimension)] + range(start, end))
    }

    /// continent | country | city | language.
    func analyticsGeography(workspaceId: String, dimension: String, start: String, end: String) async throws -> WebAnalyticsRows<WebAnalyticsRow> {
        try await client.get(analyticsPath(workspaceId, "geography"), query: [("dimension", dimension)] + range(start, end))
    }

    /// browser | os | device.
    func analyticsTechnology(workspaceId: String, dimension: String, start: String, end: String) async throws -> WebAnalyticsRows<WebAnalyticsRow> {
        try await client.get(analyticsPath(workspaceId, "browsers-systems"), query: [("dimension", dimension)] + range(start, end))
    }

    /// top | entry | exit | new.
    func analyticsPages(workspaceId: String, kind: String, start: String, end: String) async throws -> WebAnalyticsRows<WebAnalyticsPage> {
        try await client.get(analyticsPath(workspaceId, "pages"), query: [("kind", kind)] + range(start, end))
    }

    func analyticsEvents(workspaceId: String, start: String, end: String) async throws -> WebAnalyticsRows<WebAnalyticsEvent> {
        try await client.get(analyticsPath(workspaceId, "events"), query: range(start, end))
    }
}

File: windows-native/src/Webyar.Core/Api/WebAnalytics.cs
using System.Text.Json.Serialization;

namespace Webyar.Core.Api;

// ── Website analytics (`/api/web-analytics/:workspaceId/…`), as the web's SEO → Web Analytics reads it ──
// These reports answer in camelCase, unlike the rest of the API, so every field names its key.

/// <summary>The headline numbers for a date range, and the day-by-day trend.</summary>
public sealed record WebAnalyticsOverview(
    [property: JsonPropertyName("sessions")] int? Sessions = null,
    [property: JsonPropertyName("pageviews")] int? Pageviews = null,
    [property: JsonPropertyName("avgPagesPerSession")] double? AvgPagesPerSession = null,
    [property: JsonPropertyName("uniqueVisitors")] int? UniqueVisitors = null,
    // Percentage, 0–100, of sessions with a single page view.
    [property: JsonPropertyName("bounceRate")] double? BounceRate = null,
    [property: JsonPropertyName("avgVisitDurationSeconds")] double? AvgVisitDurationSeconds = null,
    [property: JsonPropertyName("trend")] IReadOnlyList<WebAnalyticsDay>? Trend = null,
    [property: JsonPropertyName("topChannels")] IReadOnlyList<WebAnalyticsRow>? TopChannels = null,
    [property: JsonPropertyName("topPages")] IReadOnlyList<WebAnalyticsPage>? TopPages = null,
    [property: JsonPropertyName("truncated")] bool? Truncated = null);

public sealed record WebAnalyticsDay(
    // YYYY-MM-DD.
    [property: JsonPropertyName("date")] string Date,
    [property: JsonPropertyName("sessions")] int? Sessions = null,
    [property: JsonPropertyName("pageviews")] int? Pageviews = null);

/// <summary>One line of a "group by X, count sessions" report.</summary>
public sealed record WebAnalyticsRow(
    [property: JsonPropertyName("key")] string Key,
    [property: JsonPropertyName("label")] string? Label = null,
    [property: JsonPropertyName("sessions")] int? Sessions = null,
    [property: JsonPropertyName("pageviews")] int? Pageviews = null);

public sealed record WebAnalyticsPage(
    [property: JsonPropertyName("path")] string Path,
    [property: JsonPropertyName("views")] int? Views = null);

public sealed record WebAnalyticsEvent(
    [property: JsonPropertyName("eventName")] string EventName,
    [property: JsonPropertyName("count")] int? Count = null,
    [property: JsonPropertyName("uniqueSessions")] int? UniqueSessions = null,
    // 0–1: the share of the range's sessions that fired it.
    [property: JsonPropertyName("conversionRate")] double? ConversionRate = null);

/// <summary>A report's rows, and whether the server stopped short of the whole range.</summary>
public sealed record WebAnalyticsRows<TRow>(
    [property: JsonPropertyName("rows")] IReadOnlyList<TRow>? Rows = null,
    [property: JsonPropertyName("truncated")] bool? Truncated = null);

internal sealed record LiveVisitorsResponse([property: JsonPropertyName("count")] int? Count = null);

public sealed partial class WebyarApi
{
    private static string AnalyticsPath(string workspaceId, string tail) =>
        $"/api/web-analytics/{Uri.EscapeDataString(workspaceId)}/{tail}";

    private static KeyValuePair<string, string?>[] Range(string start, string end, params KeyValuePair<string, string?>[] more) =>
        [.. more, Q("startDate", start), Q("endDate", end)];

    public async Task<WebAnalyticsOverview> AnalyticsOverviewAsync(string workspaceId, string start, string end, CancellationToken ct = default) =>
        await _client.GetAsync<WebAnalyticsOverview>(AnalyticsPath(workspaceId, "overview"), Range(start, end), ct).ConfigureAwait(false) ?? new();

    /// <summary>Visitors on the site right now.</summary>
    public async Task<int> AnalyticsLiveVisitorsAsync(string workspaceId, CancellationToken ct = default) =>
        (await _client.GetAsync<LiveVisitorsResponse>(AnalyticsPath(workspaceId, "live-visitors"), ct: ct).ConfigureAwait(false))?.Count ?? 0;

    /// <summary>channel | source | campaign.</summary>
    public async Task<WebAnalyticsRows<WebAnalyticsRow>> AnalyticsTrafficSourcesAsync(string workspaceId, string dimension, string start, string end, CancellationToken ct = default) =>
        await _client.GetAsync<WebAnalyticsRows<WebAnalyticsRow>>(AnalyticsPath(workspaceId, "traffic-sources"), Range(start, end, Q("dimension", dimension)), ct).ConfigureAwait(false) ?? new();

    /// <summary>continent | country | city | language.</summary>
    public async Task<WebAnalyticsRows<WebAnalyticsRow>> AnalyticsGeographyAsync(string workspaceId, string dimension, string start, string end, CancellationToken ct = default) =>
        await _client.GetAsync<WebAnalyticsRows<WebAnalyticsRow>>(AnalyticsPath(workspaceId, "geography"), Range(start, end, Q("dimension", dimension)), ct).ConfigureAwait(false) ?? new();

    /// <summary>browser | os | device.</summary>
    public async Task<WebAnalyticsRows<WebAnalyticsRow>> AnalyticsTechnologyAsync(string workspaceId, string dimension, string start, string end, CancellationToken ct = default) =>
        await _client.GetAsync<WebAnalyticsRows<WebAnalyticsRow>>(AnalyticsPath(workspaceId, "browsers-systems"), Range(start, end, Q("dimension", dimension)), ct).ConfigureAwait(false) ?? new();

    /// <summary>top | entry | exit | new.</summary>
    public async Task<WebAnalyticsRows<WebAnalyticsPage>> AnalyticsPagesAsync(string workspaceId, string kind, string start, string end, CancellationToken ct = default) =>
        await _client.GetAsync<WebAnalyticsRows<WebAnalyticsPage>>(AnalyticsPath(workspaceId, "pages"), Range(start, end, Q("kind", kind)), ct).ConfigureAwait(false) ?? new();

    public async Task<WebAnalyticsRows<WebAnalyticsEvent>> AnalyticsEventsAsync(string workspaceId, string start, string end, CancellationToken ct = default) =>
        await _client.GetAsync<WebAnalyticsRows<WebAnalyticsEvent>>(AnalyticsPath(workspaceId, "events"), Range(start, end), ct).ConfigureAwait(false) ?? new();
}
